### PR TITLE
[WIP] Minimax search function

### PR DESCRIPTION
**Pull request by a1shadows**
Issue #9

### CODE REVIEW

**Comment by a1shadows** [comment]
@BusyByte Hey, could you please resolve the conflicts and review the pull request. Thanks.

File: core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala]: I'm not a big fan of single letter parametric names when they actually mean things.  I'd use upper case and use the full name like `STATE`, `ACTION`, and `PERCEPT` rather than `S`, `A`, and `P` respectively.

File: core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala]: see above

File: core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala]: It appears that this may be tail recursive.  please add a `@tailrec` annotation to the method to checkit is.

File: core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala]: Rather than using `Action` should be using the parametric type `ACTION` (you currently have as `A`)

File: core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala]: You'll not want to use `NoAction` but instead pass into minMaxDecision a parameter of type `ACTION` (you currently have as A) 

eg
`def minMaxDecision[S, A, P](g: Game[S][A][P], noAction: A)`

**Review by BusyByte — CHANGES_REQUESTED** [review]
Some changes needed but looks good overall.
Can you add a test specification (specs2) and/or a scalacheck (property based check) to prove it works?
If there are examples from the book sometimes I'll use those to prove it works.

**Comment by BusyByte** [comment]
@a1shadows sorry something wasn't quite right with my resolution of the merge conflicts, hopefully you can fix it up

**Comment by BusyByte** [comment]
@a1shadows let me know if you have any questions on the above or need help

**Comment by a1shadows** [comment]
@BusyByte Have tried addressing all the requested changes. Am working on testing the code. Could you please review the changes?

**Review by BusyByte — APPROVED** [review]
@a1shadows this looks good but still missing a test specification/property based test.

--edit never mind above, I just saw you were working on the testing code still

**Comment by a1shadows** [comment]
> @a1shadows this looks good but still missing a test specification/property based test

I'll do the test code. I have my mid-terms till the 28th. Just need a little bit of time. Thanks for your constant guidance.

**Comment by a1shadows** [comment]
@BusyByte I've written a test for the function, but I'm having some difficulty coming up with more tests. Could you please review the new code and suggest how I should proceed with the tests?

**Comment by BusyByte** [comment]
@a1shadows fyi travis has broken, I've merge a fix to master for it now:
https://github.com/aimacode/aima-scala/pull/72

you'll need to merge master in order to get travis to run your tests

File: core/src/main/scala/aima/core/search/problems/TwoPlyGame.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/problems/TwoPlyGame.scala]: if there can only be 2 players why not code it that way;
```scala
sealed abstract class Player {
def name: String
}

final case class Player1(name: String) extends Player
final case class Player2(name: String) extends Player
```

File: core/src/main/scala/aima/core/search/problems/TwoPlyGame.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/problems/TwoPlyGame.scala]: Seems like this might go away with changes I outlined below to Action.

**Review by BusyByte — APPROVED** [review]

**Comment by BusyByte** [comment]
@a1shadows might be worth going through a full game and verifying each player's actions are what you expect and not just the first one and verify you get NoAction at the end.
I left some suggestions above but this seems pretty good so far.

**Comment by BusyByte** [comment]
@a1shadows also since this is `adversarial` maybe verify who should `win` and maybe a comment as to why

**Comment by a1shadows** [comment]
I'm getting a ton of errors on the code when I'm running it on the sbt which weren't there earlier. I can't understand them.
They are all of this kind:

aima-scala\core\src\main\scala\aima\core\agent\Actuator.scala:8:16: type parameter Action defined in trait Actuator shadows trait Action defined in package agent. You may want to rename your type parameter, or possibly remove it.

Could you point me to what's going wrong, please?

**Comment by a1shadows** [comment]
Also, would it be possible to set up a gitter channel to discuss the repository? For general queries and the like?

**Comment by BusyByte** [comment]
@a1shadows I've created a gitter channel, you should get an invitation,  I think the error you are seeing is because of a name collision.  Maybe try upper case `A` in your type parameter or upper case `ACTION`
for example.
`trait Foo[A]` or `trait Bar[ACTION]`

**Comment by BusyByte** [comment]
@a1shadows maybe you need to do a sbt clean ` sbt ";clean;test:compile"` after you merge master into your branch.
I don't think there is a `trait Action defined in package agent` in package agent anymore looking at master.

**Comment by BusyByte** [comment]
@a1shadows this still says `WIP` in the title.  Is there anything else you'd like to do before merging this?

**Comment by a1shadows** [comment]
Yeah. I'd like to work a bit on it.

**Comment by a1shadows** [comment]
@BusyByte Hey, I was attempting to rewrite the code to give an optimal state for the game at every level as opposed to my earlier code that only gave the next best move, but I just realized that the pseudocode only provides an algorithm for the next best move with the assumption that the "MAX" player is making the move. I am unsure how to proceed from here on. Should I try to write the code for the complete algorithm that can calculate moves for both "MIN" and "MAX" players or should I restrict myself to what the book has provided?

File: core/src/main/scala/aima/core/search/problems/TwoPlyGame.scala
Comment (a1shadows) [reply core/src/main/scala/aima/core/search/problems/TwoPlyGame.scala]: Could you please give an example of the format in which I'm supposed to give the comment about the function?

**Comment by BusyByte** [comment]
@a1shadows Looking at the [test for the Java version](https://github.com/aimacode/aima-java/blob/AIMA4e/test/src/test/java/aima/test/unit/search/adversarial/SearchForAdversarialActionFunctionTest.java#L67) it looks like it should handle both players.

Looking at the [java implementation](https://github.com/aimacode/aima-java/blob/AIMA4e/core/src/main/java/aima/core/search/basic/adversarial/MinimaxDecision.java#L118)  it just determines which player's turn it is based on the state.  So when you call the search function you give it a state so it should just handle it naturally.

I think it selects the max utility of the minimum utility assuming the next player will pick the min.  So it doesn't look like the algorithm needs changed to handle both users.  

I think your spec may need to change to handle both users.
It looks like the java spec is implementing aima 3e Fig 5.2

Sorry for the delay in response.  Work has been pretty hectic lately.

**Comment by BusyByte** [comment]
@a1shadows is this good to go?  If you are satisfied I can merge it.

**Comment by a1shadows** [comment]
Ahh. Sorry. I've been busy with work. I'll wrap it up within this weekend.

**Comment by a1shadows** [comment]
Still ironing a few things out. I'll let you know when I'm comfortable with the changes. Is that okay?

**Comment by BusyByte** [comment]
@a1shadows yeah, that's fine but please prioritize this as there are some more PRs coming through and I don't want you to have a bunch of merge conflicts to work through since your PR predates them.

**Comment by a1shadows** [comment]
I'm satisfied with my changes. Can you review and merge?

File: core/src/main/scala/aima/core/search/adversarial/Game.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/adversarial/Game.scala]: Really this should be `typeclass`

File: core/src/main/scala/aima/core/search/adversarial/Game.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/adversarial/Game.scala]: I don't really like having traits with no defined behavior.
I'd just as soon keep these in the definition of the given game so they can be sealed and compile time checked to be total matched.

**Review by BusyByte — APPROVED** [review]
Some things I'd change but this looks good.

File: core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala
Comment (BusyByte) [inline core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala:48]: @a1shadows I was looking at this when trying to implement alpha beta search and have a question.  Why are you passing in the `g.initialState` rather than `state` which is passed into the function?  this does not seem to match the algorithm.

**Comment by a1shadows** [comment]
I decided to not leave that decision up to the caller because the correct
behavior will change according to the state for which the decision will be
made. For some of the nodes, minMaxValue has to be called for the correct
decision and not maxMindecision. Would me like me to change the logic?
Sorry for the late reply.

**Pull request merged**
